norfair/video.py

```diff
@@ -146,7 +146,7 @@ def write(self, frame: np.array) -> int:
 
     def show(self, frame: np.array, downsample_ratio: int = 1) -> int:
         # Resize to lower resolution for faster streaming over slow connections
-        if downsample_ratio is not None:
+        if downsample_ratio not in [0, 1, None]:
```

Comment:
I think more than not 0, 1, or None, it should be not None, not 1, and larger than 0, so as to avoid negative numbers.

Reply:
I agree, I made a slight change to only allow positive values ​​greater than 1.

https://github.com/tryolabs/norfair/pull/15/commits/d58c8b3150d9c749163be2197b2803c1a58657d4

Reply:
Related to None, it is possible for someone to assign the value None to the argument, if the control is not incorporated, the comparison fails.

Reply:
Thanks for the changes. I think it should actually be `larger than 0` and also `not equal to 1` so as to allow for people to upsample the image if they need to. We should change the type in the function signature to `float` to allow for this. Regarding `None` still failing, that should be mitigated by python's type hints, we can't really check for everything that someone might decide to place in that argument.

Reply:
The problem with typing mentioned in its documentation
"The Python runtime does not enforce function and variable type annotations."
It is a hint to be used at inspection time (IDE, lint), it is good for internal code but when exposed as an api, I think errors may occur (but it is true, each evaluation is an expense in CPU, it is a design decision).

For these cases, if I could change the way these routines work that are always involved in an iteration, it is to change the sets to a separate method where it is validated and assigned to preset internal values and then directly used in runtime the last assigned and valid value.
This minimizes CPU usage in comparisons and validations in each iteration.
It also doesn't make sense that they are continuously changing the scaling value every frame.
In case someone wants to, they could repeatedly call the set, and do it.

Related to allowing scaling and allowing floating, I think the idea is a good one.
I think you could evaluate using fx and fy (scale factor), to be able to handle the values ​​and not be calculating the width and height pixel values ​​each time.

Currently this PR was a suggestion, a warning about mainly the value 1, which calls resize without having to do it by default.
I can work on making the change to float and use fx, fy, and try to minimize the evaluation in each iteration (on this or another PR).
I am waiting for information, what action I do over this PR.

Reply:
Gotcha. I get what you mean about python's type hints not being a robust way to check for users using the correct types when calling functions. The thing is that python is a dynamically typed language, and the only way we could robustly check for this is to add exhaustive checks to every function input, to check that the user inserted the correct types. This has a runtime and also code base maintenance/readability cost.

The usual workflow for someone using a python function from a library is for them to check the documentation, or if they are using a good IDE, tab complete and get the function signature, both things we currently support as we have good documentation and type hinting on all functions. I think we'll expect this from users of this library and not manually check types on all functions.

Regarding None, what I was saying previously was that this function expects a float, and the documentation and the type hinting reflect that. We could of course check that the user didn't insert a None, but there is nothing special about None here, we could also check that the user didn't insert a List, or a Dict, or even check that the user did actually insert a Float explicitly, but we'd have to do this through all of our codebase which doesn't make too much sense.

I think that just checking for `downsample_ratio > 0` and raising an error if false, and check for `downsample_ratio != 1` and not performing the downsampling operation if thats the case, we'd be set.

Again thanks for the PRs!

Reply:
Following the idea, it does not make sense to incorporate error controls, if someone sends a value to zero, the assert of the resize method already performs that control, so an extra evaluation could be delegated.
Leaving only the downsample_ratio! = 1 evaluation would obtain the lowest computation cost in each iteration.
Regarding the change of data type, I am going to make a proposal for a change in the handling of that data type and the resize mechanism used in another PR. If it's OK for you.

Reply:
Yeah no problem, let's follow up on the other PR.